include/cantera/base/Units.h

```diff
@@ -77,6 +81,50 @@ class Units
 };
 
 
+//! Unit aggregation utility
+/*!
+ *  Provides functions for updating and calculating effective units from a stack
+ *  of unit-exponent pairs. Matching units are aggregated, where a standard unit
+ *  simplifies access when joining exponents. The utility is used in the context
+ *  of effective reaction rate units.
+ */
+struct UnitStack
+{
+    UnitStack(const Units& standardUnits) {
+        stack.reserve(2); // covers memory requirements for most applications
+        stack.emplace_back(standardUnits, 0.);
+    }
+
+    //! Alternative constructor allows for direct assignment of vector
+    UnitStack(std::initializer_list<std::pair<Units, double>> units)
+        : stack(units) {}
+
+    //! Size of UnitStack
+    int size() const { return stack.size(); }
+
+    //! Get standard unit used by UnitStack
+    Units standardUnits() const;
+
+    //! Set standard units
+    void setStandardUnits(Units& standardUnits);
+
+    //! Effective exponent of standard unit
+    double standardExponent() const;
+
+    //! Join exponent of standard units
+    void join(double exponent);
+
+    //! Update exponent of item with matching units; if it does not exist,
+    //! add unit-exponent pair at end of stack
+    void update(const Units& units, double exponent);
+
+    //! Calculate product of units-exponent stack
+    Units product() const;
+
+    std::vector<std::pair<Units, double>> stack; //!< Stack uses vector of pairs
+};
```

Comment:
This could be protected and named `m_stack`, right? I don't think it's accessed publically anywhere.

Reply:
While it is currently not accessed publicly anywhere at the moment, I declared this class as a `struct` as the intent was to keep things simple. In addition to marking it as experimental, I am also adding the `@internal` parameter to the docstring.